gloo/benchmark/runner.cc

```diff
@@ -307,5 +346,44 @@ template void Runner::run(BenchmarkFn<float>& fn, int n);
 template void Runner::run(BenchmarkFn<float16>& fn);
 template void Runner::run(BenchmarkFn<float16>& fn, int n);
 
+RunnerThread::RunnerThread() : stop_(false), job_(nullptr) {
+  thread_ = std::thread(&RunnerThread::spawn, this);
+}
+
+RunnerThread::~RunnerThread() {
+  mutex_.lock();
+  stop_ = true;
+  mutex_.unlock();
+  cond_.notify_one();
+  thread_.join();
+}
+
+void RunnerThread::run(RunnerJob* job) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  job_ = job;
+  cond_.notify_one();
+}
+
+void RunnerThread::spawn() {
+  std::unique_lock<std::mutex> lock(mutex_);
+  while (!stop_) {
+    while (job_ == nullptr) {
+      cond_.wait(lock);
+      if (stop_) {
+        return;
+      }
+    }
```

Comment:
Maybe copy job_ to a temporary location so you can release the lock here and allow for faster shutdown? This assumes that the job is only being run by one thread at a time (I think that's the case) since samples_ is not thread safe.